ENH: added vector operators: divergence, curl and laplacian

antlarac opened the pull request:
Hello, yesterday I tried to upload this change but I messed up (sorry, my first time on github). I think I got it right now. In this commit I have included three new functions to /numpy/lib/function_base.py, namely divergence, curl and laplacian. I have been missing these functions when using numpy myself.

Thank you,

Antonio

Commit:
ENH: added vector operators: divergence, curl and laplacian

antlarac commented:
I am trying to notify to the email list about this, as suggested in the workflow, but I can't figure out how to send a message to a specific list. Is it done in Gmane? Any help on this would be greatly appreciated.
Thanks, 
Antonio

rkern commented:
Subscribe here: https://mail.scipy.org/mailman/listinfo/numpy-discussion

Or use Gmane: http://dir.gmane.org/gmane.comp.python.numeric.general

jaimefrio commented:
Your formatting seems to be way off, please:

 * use 4 space indentation throughout,
 * break long lines at 79 characters, and
 * follow the example of any other docstring for the proper vertical and horizontal spacing there.

Commit:
ENH: added vector operators: divergence, curl and laplacian

antlarac commented:
Ahh, I'm sorry for that, I didn't realize... I think everything is ok now.
Thanks for the comment!

argriffing commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
+    Return the laplacian of an N-dimensional array.
+
+    Computes the laplacian of an N-dimensional numpy array, as a
+    concatenation of the numpy functions gradient and divergence. 
```
Instead of 'concatenation' do you mean 'composition' here? Or this could be jargon that is unfamiliar to me, for example the composition of functions might be written as the concatenation of the symbols representing the functions. And what is the meaning of the `_s` in `laplace_s`?

antlarac commented:
Hello  argriffing,
Yes, probably composition is more appropriate here, I just mean applying gradient to the input argument, and then applying divergence to the result of that gradient.
As for the _s, it means the "scalar laplacian", since laplacian operator also exists for vector fields. I was planning to implement it also. In principle it is just applying first the divergence and then the gradient, but I want to make sure first, it is not so straightforward for vector fields (https://en.wikipedia.org/wiki/Vector_Laplacian), but I'm pretty sure that in this case, that everything is done in "cartesian coordinates" it's just that. So I was thinking about also creating another function for this case, called laplace_v, but of course the names can be changed to more appropriate ones. For example I know that in Matlab they call "del2" to the laplacian operator.
Thanks,

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
+    
+    Parameters
+    ----------  
+    v : list of numpy arrays
```
Wouldn't it make more sense to make this an array-like? Here, one can choose which dimension to be that of the axes; I think the first one is fine.  In that case, one could just do
```
v = np.asanyarray(v)
```
and it would even cover the case of having a list of numpy arrays. Another advantage would be that the tests for the sizes to be the same could be omitted, since that would be done implicitly already.

antlarac replied:
Hi,
Thanks for your suggestion. By reviewing the size test, I have realized that they are incomplete right now, they only check that every component has the same number of dimensions, but not the same size. A quick example:
>>> a = np.array([[1,2],[3,4]])
>>> b = np.array([[5,6,7],[8,9,10],[11,12,13]])
>>> c=[a,b]
>>> d = np.asanyarray(c) 
>>> np.shape(d[0])
(2,2)
>>> np.shape(d[1])
(3,3)
This should fail the tests for size. But the tests right now would pass since:
>>> len(d[0].shape)
2
>>> len(d[1].shape)
2

I'm going to fix this and will implement your suggestion as well. 
Thank you

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
+    
+    
+    
+    axes = tuple(range(N))
```
Is this a mistake? Given this, why is there a need to normalize the axes below?

antlarac replied:
Yes, you are right, I borrowed some code from np.gradient and forgot to remove the normalization below. I will do it now, thanks.

seberg commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
     else:
         return outvals
 
+def divergence(v,*varargs,**kwargs):
```
Some more general comments -- did not make up my mind about everything, so some of it is more for discussion:
  * I know varargs is copied from gradient, but I am frankly not a fan, I think I would prefer varargs to be a single sequence argument.
  * There seems to be quite a lot of duplication with gradient, maybe we can just call gradient (possible add some arguments to gradient, such as `axis=None` to pick which derivatives to calculate or some such; that would also in some sense allow the stacked matrix kind of logic)

antlarac replied:
Hello seberg,
You are right, it makes more sense to call gradient as needed. I thought of this but for some reason I also thought that each function should be "stand alone" (which I didn't do in laplace!). It will certainly be easier to understand (for me too!) in this way. I will work on keeping what is new, and have the rest substituted for the appropriate calls to gradient. Yes, all the varargs are uncofortable to me too, In this way (calling gradient), this "freedom" will be removed when calling divergence, curl and laplacian, since they not make much sense for these functions. 
Thanks for your suggestions

seberg commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
+        i.e. `dx`, `dy`, `dz`, ... Default distance: 1.
+        single scalar specifies sample distance for all dimensions.
+        if `axis` is given, the number of varargs must equal the number of axes.
+    edge_order : {1, 2}, optional
```
This seems to not actually be implemented/make a difference currently. Which also points to an unfortunate huge chunk of work, but it can wait for a while, just needs patience. And that is, we need a lot of tests after hashing out what we want.

seberg commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
+    >>> plt.show()
+    """
+    
+    l = np.divergence(np.gradient(f))
```
While I know this is mathematically sound, I am not sure that this is numerically very good or efficient?

I don't mind having a reasonable good and simple default laplacian in numpy, but I have a bit doubts that this is a good approach. Unfortunatly the numpy correlation function does not have a "mirror" mode, which would make a (typical?) stencil implementation very easy (more similar to the scipy one).

Commit:
ENH: added vector operators: divergence, curl and laplacian

shoyer commented:
One thing that I would find helpful is to include the definition of the relevant operators (written using partial derivatives) in docstrings for each function.

Commit:
ENH: added vector operators: divergence, curl and laplacian

antlarac commented:
Following some suggestions, I have eliminated duplicated code, by calling gradient with the appropriate "axis" parameter every time. Now everything looks cleaner and easier to understand. The definitions (using derivative notation) are included in the docstrings, in the return section

shoyer commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+    for i, v_c in enumerate(v):
+        v_c = np.asanyarray(v_c)
+        N[i] = len(v_c.shape) # Get number of dimensions from every component
+    if False in [N[0] == N[i] for i in range(len(v))]: 
```
use `if not all(...)` instead of `if False in [...]`

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+     [ 0.  0.  0. ...,  0.  0.  0.]
+     [ 0.  0.  0. ...,  0.  0.  0.]
+    """
+    N = [0]*len(v)
```
My sense is that these pre-calculation shape checks are not necessary: you can just rely on the addition in outvals failing if the shapes do not match.  But if you really feel it is necessary, all that would be required is
```
if len(set(np.shape(item) for item in v)) > 1:
    raise ...
```

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+	# If all vector field components are the same shape,
+	#N becomes the number of dimensions
+    
+    axes = tuple(range(N))
```
This seems overly long as well. Doesn't the below cover it?
```
outvals = gradient(v[0], *varargs, axis=0, edge_order=0)
for axis in range(1, len(v)):
    outvals += gradient(v[axis], *varargs, axis=axis, edge_order=0)
return outvals
```

mhvk replied:
p.s. Is there any reason not just to have `edge_order` as a keyword argument and pass it on?

mhvk replied:
Neat.  Though I fear it fails the Zen of Python...

Actually, I think python's `sum` is smart enough to do this intrinsically, i.e., one would just do
```
outvals = sum(gradient(v[axis], *varargs, axis=axis, edge_order=0)
              for axis in range(len(v)))
```
This definitely beats everything else in readability, but, doing a quick test, is quite a bit slower than either of our approaches (which both are faster than first making a zero-filled array for large sizes)

shoyer replied:
Indeed. The problem with `sum` is that it doesn't do in-place arithmetic on the result, e.g.,
```
class A(object):
    def __iadd__(self, other):
        print('iadd with %r' % other)
        return self

    def __add__(self, other):
        print('add with %r' % other)
        return self

    def __radd__(self, other):
        print('radd with %r' % other)
        return self

sum([A()] * 4)
# radd with 0
# add with <__main__.A object at 0x107590748>
# add with <__main__.A object at 0x107590748>
# add with <__main__.A object at 0x107590748>
```

Maybe defining `inplace_sum` would be cleaner?
```python
def inplace_sum(items, start=0):
    total = start
    for item in items:
        total += item
    return total
```

shoyer replied:
Actually, in this case we want the result to be added in-place on the first object to avoid an unnecessary memory allocation. So really, we need something like this:
```python
def inplace_sum(items):
    it = iter(items)
    total = next(it)
    for item in it:
        total += item
    return total
```

mhvk replied:
Yes, that makes sense -- I can see how for a generic `sum` you cannot rely on inplace addition doing what you'd like. You solution also makes sense, and is similarly fast to our other options. Still, probably getting a bit off-topic here -- I think for the divergence I would just stick with doing it explicitly as in my original suggestion.

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+    >>> plt.colorbar()
+    >>> plt.show()
+    """
+    N = [0]*len(v)
```
See comment on divergence above.

mhvk replied:
ps. here of course you do need to check that `len(v) == 3`.

antlarac replied:
Do you mean len(v) == 3 in divergence? In principle I was intending to create a "generalized" divergence, for any number of dimensions, following the example of np.gradient, that can be applied to N-dimensional arrays, and the "return vector" would have N components. Please, correct me if I am wrong. However, for the case of curl, three dimensions are needed, not more nor fewer.

mhvk replied:
No, for divergence I felt you either should omit all sanity checks, since if the shapes did not match the calculation of the divergence would fail anyway, or just do the very simple one I gave, since that covers what you had anyway.
My note was just to say that for curl, in addition to the (possible) check that shapes are identical, you do need to check that `v` has length 3 (as you say too).

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+        # If all vector field components are the same shape,
+        #N becomes the number of dimensions
+
+    outvals=[np.zeros(np.shape(v_c))] * 3 #Initialize output vector field
```
This can also be shortened substantially
```
# Calculate dAz/ dy - dAy / dz, dAx / dz - dAz / dx, dAy / dx - dAx / dy.
outvals = [gradient(v[(axis + 2) % 3], *varargs, axis=(axis + 1) % 3, edge_order=0) -
           gradient(v[(axis + 1) % 3], *varargs, axis=(axis + 2) % 3, edge_order=0)
           for axis in range(3)]
return outvals
```
(Note that I don't call `np.gradient` since the function `gradient` is defined locally anyway)

antlarac replied:
Hello everyone,
Thank you for all your suggestions, they will definitely improve the functions. I will start implementing them, I guess it will take me a couple of days to review everything and make some profiles to filter the most efficient ones. By the way, in my last commit I force edge_order to 0 when calling gradient because if it is 1 something wrong happens in the edges, and the examples don't return arrays of zeros (in the edges there are 0.5 and -0.5). With edge_order = 0 it does not happen.
Thanks again

Commit:
ENH: added vector operators: divergence, curl and laplacian

Commit:
ENH: added vector operators: divergence, curl and laplacian

antlarac commented:
I have incorporated your suggestions. Everything looks much better and easier now!

antlarac commented:
Hi again,
Sorry for asking something so simple, this is my first contribution to GitHub. If everyone agrees with the last version, what is the next step? Do I have to write to the newsletter again?

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,201 @@ def gradient(f, *varargs, **kwargs):
+    
+    N=len(v)
+    axes = tuple(range(N))
+    out = inplace_sum([gradient(v[ax],*varargs,axis=ax,edge_order=0)
```
You'd have to pass the argument as an iterator here, otherwise the list with all the arrays still gets created. I think all you need to do is to remove the square brackets.

While you are at it, also follow PEP8 more closely: in particular, spaces around `=` (line 1388), spaces after `,`, and the `for ax in axes` should align with the first character after `inplace_sum(`
(If you use emacs, these kinds of checks can be done on-the-fly: http://docs.astropy.org/en/latest/development/codeguide_emacs.html)

antlarac replied:
Hello mhvk, 
Thank you for your comment. I'm not sure if I understand what you mean. Do you propose to not create a list (in the example, "c") to pass to the function, but instead do something like a generator? In order to save memory? In that case, I think something inside the function should be changed. For example, if a list is not passed, the N = len(v) would no longer make sense, and some more changes would be necessary.
Thanks

mhvk replied:
No, I just meant to pass on a generator rather than a list to `inplace_sum`, not to the whole function. Here, on line 1390, you are creating a list implicitly, by using the `[` and `]` at the ends; these should be removed.

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,201 @@ def gradient(f, *varargs, **kwargs):
+    if len(v) != 3:
+        raise ValueError("Enter a list of 3 arrays")
+
+    outvals = [gradient(v[(ax+2)%3], *varargs, axis=(ax+1)%3, edge_order=0)-
```
PEP8 would have spaces around `%`, and in front of `-`.

mhvk commented:
@antlarac - I added a few small further comments.  One more one is that all code should have tests that check that they work as expected (and more importantly, *continue* to work as expected when someone else starts fiddling with the code). Tests for this module seem to be in `numpy/lib/tests/test_function_base.py` -- so that would seem the logical place to add some more.

antlarac commented:
Thank you for your answers, I will work on the tests now

charris commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,201 @@ def gradient(f, *varargs, **kwargs):
     else:
         return outvals
 
+def inplace_sum(items): 
```
If a private function, should start with an underscore

charris commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,201 @@ def gradient(f, *varargs, **kwargs):
+    Returns the sum of the elements of a list. 
+    Used in the follwing functions:
+    divergence,
+    """
```
Blank line before.

charris commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1320,201 @@ def gradient(f, *varargs, **kwargs):
+    order accurate one-sides (forward or backwards) differences at the
+    boundaries. The returned gradient hence has the same shape as the
+    input array.
+    
```
Should have `.. versionadded:: 1.11.0` here. Grep for examples.

charris commented:
The new functions need to be added to the `__all__` list and a note should be added in `doc/release/1.11.0-notes.rst`. At some point the commits also need squashing.

antlarac commented:
Thank you for all the comments on formatting, I am fixing everything :)

Commit:
ENH: added vector operators: divergence, curl and laplacian

Commit:
ENH: added vector operators: divergence, curl and laplacian

antlarac commented:
Hello again,
I finally added some tests and fixed some other things. Also added the description of new functions in doc/release/1.11.0-notes.rst.
Can anyone think of any other way to improve this contribution?
Thank you :)

seberg commented:
I am sorry, it seems this is difficult to get moving :(.

Anyway, as I mentioned before.... I am going to ask you or someone else to convince me that this is actually a **reasonable** way to calculate the laplacian numerically, or to scrap the whole function. I don't doubt that it is in principle correct, but I am *very* dubious about it. For one thing, it must have a rather large stencil (In my world laplacians are numerically calculated with stencils), but is the boundary really quite correct for that stencil, etc.? Note for example:
```
In [119]: r = np.sin(np.linspace(0, np.pi, 500))[None].repeat(500, axis=0)

In [120]: np.laplace(r).sum()
Out[120]: -6.2959224292199103

In [121]: scipy.ndimage.laplace(r).sum()
Out[121]: 5.0480886831794081e-13
```

And because I am dubious about the laplace, I think I need some reassurance that the other stuff is numerically reasonable, I don't really doubt it, but if you have other softwares that calculate these values did you compare the results?

--

There is at least one code problem  that might be a bit more serious (older than your stuff really) though probably a simple fix. `edge_order=0` is used, which is not even supported and should raise an error. Since it is used though, what does it mean?

--

Last point, I am still a bit unsure about the signature. Ignoring `gradient` which I am sure is ancient, might not a signature such as (with or without axis) be convenient?:
```
def divergence(arr, axis=-1, diffs=None):
    """
    Parameters
    ---------------
    arr : ndarray
         Data for which to calculate divergence
    axis : int
         Axis signifying the vector dimension.
    """
```
Of course in some sense this is less efficient if you happen to have your data in list-form, but it seems to me to be the more common form in numpy when considering linalg. Just thinking, but I think before the end I would like to poll the mailing list about the API again unless some more chip in here.

antlarac commented:
ok, thanks for your comment, let me check it.

antlarac commented:
I'm sorry this is taking me so long, these days I can't dedicate much time...

One alternative for the laplacian would be to apply gradient twice in each direction, and sum the results. I am concerned with what would happen in the edges, but this point is always trickier anyway...

Also, I have found the source code for octave's "del2" function. Using the proposed np.laplace for several functions, the results are qualitatively equal for many tests with arbitrary functions, but there is a quantitative difference (for some reason, the del2 function in octave and matlab is 1/4 times the actual laplacian, but the octave results actually are slightly smaller than that, compared to the proposed np.laplace). Also, I have found some problems with 1D functions, I will look deeper into that.

I will check the first option (using gradient twice), as well as trying to implement a python version of octave's del2 function. If anyone thinks one of these approaches is not reasonable for numpy, please let me know. For the sake of curiosity, next I copy the octave del2 function:

Thanks for your help

## Copyright (C) 2000-2013 Kai Habel
## Copyright (C) 2007  David Bateman
##
## This file is part of Octave.
##
## Octave is free software; you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or (at
## your option) any later version.
##
## Octave is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Octave; see the file COPYING.  If not, see
## <http://www.gnu.org/licenses/>.

## -*- texinfo -*-
## @deftypefn  {Function File} {@var{d} =} del2 (@var{M})
## @deftypefnx {Function File} {@var{d} =} del2 (@var{M}, @var{h})
## @deftypefnx {Function File} {@var{d} =} del2 (@var{M}, @var{dx}, @var{dy}, @dots{})
##
## Calculate the discrete Laplace
## @tex
## operator $( \nabla^2 )$.
## @end tex
## @ifnottex
## operator.
## @end ifnottex
## For a 2-dimensional matrix @var{M} this is defined as
## @tex
## $$d = {1 \over 4} \left( {d^2 \over dx^2} M(x,y) + {d^2 \over dy^2} M(x,y) \right)$$
## @end tex
## @ifnottex
##
## @example
## @group
##       1    / d^2            d^2         \
## D  = --- * | ---  M(x,y) +  ---  M(x,y) |
##       4    \ dx^2           dy^2        /
## @end group
## @end example
##
## @end ifnottex
## For N-dimensional arrays the sum in parentheses is expanded to include second
## derivatives over the additional higher dimensions.
##
## The spacing between evaluation points may be defined by @var{h}, which is a
## scalar defining the equidistant spacing in all dimensions.  Alternatively,
## the spacing in each dimension may be defined separately by @var{dx},
## @var{dy}, etc.  A scalar spacing argument defines equidistant spacing,
## whereas a vector argument can be used to specify variable spacing.  The
## length of the spacing vectors must match the respective dimension of
## @var{M}.  The default spacing value is 1.
##
## At least 3 data points are needed for each dimension.  Boundary points are
## calculated from the linear extrapolation of interior points.
##
## @seealso{gradient, diff}
## @end deftypefn

## Author:  Kai Habel <kai.habel@gmx.de>

function D = del2 (M, varargin)

  if (nargin < 1)
    print_usage ();
  endif

  nd = ndims (M);
  sz = size (M);
  dx = cell (1, nd);
  if (nargin == 2 || nargin == 1)
    if (nargin == 1)
      h = 1;
    else
      h = varargin{1};
    endif
    for i = 1 : nd
      if (isscalar (h))
        dx{i} = h * ones (sz (i), 1);
      else
        if (length (h) == sz (i))
          dx{i} = diff (h)(:);
        else
          error ("del2: dimensionality mismatch in %d-th spacing vector", i);
        endif
      endif
    endfor
  elseif (nargin - 1 == nd)
    ## Reverse dx{1} and dx{2} as the X-dim is the 2nd dim of the ND array
    tmp = varargin{1};
    varargin{1} = varargin{2};
    varargin{2} = tmp;

    for i = 1 : nd
      if (isscalar (varargin{i}))
        dx{i} = varargin{i} * ones (sz (i), 1);
      else
        if (length (varargin{i}) == sz (i))
          dx{i} = diff (varargin{i})(:);
        else
          error ("del2: dimensionality mismatch in %d-th spacing vector", i);
        endif
      endif
    endfor
  else
    print_usage ();
  endif

  idx = cell (1, nd);
  for i = 1: nd
    idx{i} = ":";
  endfor

  D = zeros (sz);
  for i = 1: nd
    if (sz(i) >= 3)
      DD = zeros (sz);
      idx1 = idx2 = idx3 = idx;

      ## interior points
      idx1{i} = 1 : sz(i) - 2;
      idx2{i} = 2 : sz(i) - 1;
      idx3{i} = 3 : sz(i);
      szi = sz;
      szi (i) = 1;

      h1 = repmat (shiftdim (dx{i}(1 : sz(i) - 2), 1 - i), szi);
      h2 = repmat (shiftdim (dx{i}(2 : sz(i) - 1), 1 - i), szi);
      DD(idx2{:}) = ((M(idx1{:}) - M(idx2{:})) ./ h1 + ...
                     (M(idx3{:}) - M(idx2{:})) ./ h2) ./ (h1 + h2);

      ## left and right boundary
      if (sz(i) == 3)
        DD(idx1{:}) = DD(idx3{:}) = DD(idx2{:});
      else
        idx1{i} = 1;
        idx2{i} = 2;
        idx3{i} = 3;
        DD(idx1{:}) = (dx{i}(1) + dx{i}(2)) / dx{i}(2) * DD (idx2{:}) - ...
            dx{i}(1) / dx{i}(2) * DD (idx3{:});

        idx1{i} = sz(i);
        idx2{i} = sz(i) - 1;
        idx3{i} = sz(i) - 2;
        DD(idx1{:}) =  (dx{i}(sz(i) - 1) + dx{i}(sz(i) - 2)) / ...
            dx{i}(sz(i) - 2) * DD (idx2{:}) - ...
            dx{i}(sz(i) - 1) / dx{i}(sz(i) - 2) * DD (idx3{:});
      endif

      D += DD;
    endif
  endfor

  D = D ./ nd;
endfunction

charris commented:
@antlarac When you get back to this, could you also squash the commits. You can use `git rebase -i HEAD~n` for that, where `n` is the number of commits you want to work on.

antlarac commented:
Ok, I didn't know about this, thank you for the tip :)

antlarac commented:
Ok, I know this is taking too long, but I think I'll finish soon. I've been figuring out how del2 from matlab / octave works, and it definitely uses 5 point stencil for interior points, and the edge points are linearly interpolated from the two nearest neighbors of each edge. So it certainly would give different results, (especially at the edges) than my initial proposal of taking the divergence of the gradient. Thanks seberg, you were completely right at this point !!

I have also been checking the matlab versions of divergence and curl, and those use gradient for every dimension as needed. So I would say that the divergence and curl proposed in this pull request are reliable.
I hope to finish the laplacian (maybe it is better to call it del2 as in matlab/octave??) in a few days. 

I'm still not 100% sure why del2 is actually 1/4 of the value of the actual laplacian. It may have to do with how they add (h1 + h2) to calculate the second derivative (DD), but that would give a factor 1/2, I think (at least for regular grids). I will keep working on this, but of course any suggestion is welcome :)

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1320,6 +1321,214 @@ def gradient(f, *varargs, **kwargs):
+    total = next(it)
+    for item in it:
+        total += item
+    return total
```
I think this can be written `np.sum(arr, out=arr[0])`. Seems like a premature optimization to me though

seberg commented:
Since the original author deleted the github account, and I think the PR still needed quite a lot of work and probably some discussion. I will open an issue for a request to add these functions to `np.linalg` pointing to this PR since there is likely some good start and discussion here.
